cmd/gateway-gcs.go

```diff
@@ -561,266 +570,190 @@ func (l *gcsGateway) DeleteObject(bucket string, object string) error {
 func (l *gcsGateway) PutObjectPart(bucket string, key string, uploadID string, partID int, size int64, data io.Reader, md5Hex string, sha256sum string) (PartInfo, error) {
-	multipartKey := toGCSMultipartKey(key, uploadID, partID)
+	meta := gcsMultipartMetaName(uploadID)
+	object := l.client.Bucket(bucket).Object(meta)
 
-	info, err := l.PutObject(bucket, multipartKey, size, data, map[string]string{}, sha256sum)
-	return PartInfo{
-		PartNumber:   partID,
-		LastModified: info.ModTime,
-		ETag:         info.ETag,
-		Size:         info.Size,
-	}, err
-}
+	_, err := object.Attrs(l.ctx)
+	if err != nil {
+		return PartInfo{}, gcsToObjectError(traceError(err), bucket, key)
+	}
 
-// ListObjectParts returns all object parts for specified object in specified bucket
-func (l *gcsGateway) ListObjectParts(bucket string, key string, uploadID string, partNumberMarker int, maxParts int) (ListPartsInfo, error) {
-	delimiter := slashSeparator
-	prefix := pathJoin(ZZZZMinioPrefix, fmt.Sprintf("multipart-%s-%s", escape(key), uploadID))
+	var sha256Writer hash.Hash
 
-	it := l.client.Bucket(bucket).Objects(l.ctx, &storage.Query{Delimiter: delimiter, Prefix: prefix, Versions: false})
+	// Generate random ETag.
+	etag := getMD5Hash([]byte(mustGetUUID()))
 
-	isTruncated := false
+	reader := data
 
-	it.PageInfo().Token = toGCSPageToken(toGCSMultipartKey(key, uploadID, partNumberMarker))
-	it.PageInfo().MaxSize = maxParts
+	if sha256sum != "" {
+		sha256Writer = sha256.New()
+		reader = io.TeeReader(data, sha256Writer)
+	}
 
-	nextPartnumberMarker := 0
+	dataName := gcsMultipartDataName(uploadID, etag)
 
-	parts := []PartInfo{}
-	for {
-		if len(parts) >= maxParts {
-			isTruncated = true
-			break
-		}
+	object = l.client.Bucket(bucket).Object(dataName)
 
-		attrs, err := it.Next()
-		if err == iterator.Done {
-			break
-		} else if err != nil {
-			return ListPartsInfo{}, gcsToObjectError(traceError(err), bucket, prefix)
+	w := object.NewWriter(l.ctx)
+	if md5Hex != "" {
+		var err error
+		w.MD5, err = hex.DecodeString(md5Hex)
+		if err != nil {
+			return PartInfo{}, gcsToObjectError(traceError(err), bucket, key)
 		}
+	}
+	_, err = io.Copy(w, reader)
+	if err != nil {
+		return PartInfo{}, gcsToObjectError(traceError(err), bucket, key)
+	}
 
-		if attrs.Prefix != "" {
-			continue
-		}
+	err = w.Close()
+	if err != nil {
+		fmt.Println(err)
+		return PartInfo{}, gcsToObjectError(traceError(err), bucket, key)
+	}
 
-		_, _, partID, err := fromGCSMultipartKey(attrs.Name)
-		if err != nil {
-			continue
-		} else if partID == 0 {
-			// we'll ignore partID 0, it is our zero object, containing
-			// metadata
-			continue
+	if sha256sum != "" {
+		if hex.EncodeToString(sha256Writer.Sum(nil)) != sha256sum {
+			object.Delete(l.ctx)
+			return PartInfo{}, traceError(SHA256Mismatch{})
 		}
+	}
+
+	return PartInfo{
+		PartNumber:   partID,
+		ETag:         etag,
+		LastModified: time.Now().UTC(),
+		Size:         size,
+	}, err
```

Comment:
Explicit `nil` for error is more readable here. The `err` variable is "stale" at this point - it is not even the last error that could have happened in the function.